### ISSUE TITLE
Incompatible with Blender 4.2

### ISSUE THREAD

**Issue by smlpt** [mastodon-sc/mastodon-blender-view#20]
I tried to run the addon installer on a portable Blender 4.2 instance, but it threw the error:

<details><summary>Exception</summary>
<p>

```
java.lang.RuntimeException: Installation of the dependencies for the mastodon_blender_view addon failed:
Looking in links: c:\Users\pantze28\AppData\Local\Temp\tmp3jkh76fh
Requirement already satisfied: setuptools in c:\software\blender-4.2.0-windows-x64\4.2\python\lib\site-packages (63.2.0)
Requirement already satisfied: pip in c:\software\blender-4.2.0-windows-x64\4.2\python\lib\site-packages (23.2.1)
Blender 4.2.0 (hash a51f293548ad built 2024-07-16 06:29:33)

Blender quit

	at org.mastodon.blender.setup.BlenderSetupUtils.installAddon(BlenderSetupUtils.java:66)
	at org.mastodon.blender.setup.BlenderSetupController.installAddonThread(BlenderSetupController.java:126)
	at java.lang.Thread.run(Thread.java:750)
```

</p>
</details> 

Trying the same with a portable Blender 4.1 works fine.

**Comment by maarzt** [comment]
@smlpt Thank you for the reporting the issue. The Mastodon Blender View plugin installs a view packages into Blender. Here it seems the installation script fails because in Blender 4.2 "setuptools" and "pip" are already installed. I think the problem can be easily solved with a few modifications to the install script. It should test if a package is installed before trying to install it. I will look into this.

**Comment by maarzt** [comment]
I tried to reproduce the problem on my machine. But both Blender 4.2.0 and 4.2.1 works fine. I'm using linux / ubuntu so maybe it is windows specific. I don't know. But this mean it will be a lot identify the problem.

My assumption that there is a problem with already installed packages also appears to be wrong. I get the same warnings:
```
Requirement already satisfied: setuptools in c:\software\blender-4.2.0-windows-x64\4.2\python\lib\site-packages (63.2.0)
Requirement already satisfied: pip in c:\software\blender-4.2.0-windows-x64\4.2\python\lib\site-packages (23.2.1)
```
But for me the installation works fine.

**Comment by maarzt** [comment]
@smlpt Could you help me debug the problem, please. Could you try to execute this script [(install_addon.py)](https://github.com/mastodon-sc/mastodon-blender-view/blob/master/src/main/resources/blender-scripts/install_addon.py) in Blender 4.2. Something like: `> blender.exe --background --python install_addon.py` should do the trick. The interesting part is whether or not blender prints "dependencies installed" and if there are other error messages.

**Comment by maarzt** [comment]
We failed to reproduce the problem even on another Windows machine. @smlpt Could you try again. Is there something that might be different on your machine?

**Comment by smlpt** [comment]
Running the addon with mobile Blender 4.1 gives me this:

<details><summary>Details</summary>
<p>

```
Blender 4.1.1 (hash e1743a0317bc built 2024-04-15 23:33:30)
Read prefs: "C:\Users\pantze28\AppData\Roaming\Blender Foundation\Blender\4.1\config\userpref.blend"
Looking in links: c:\Users\pantze28\AppData\Local\Temp\tmpygd8oj__
Requirement already satisfied: setuptools in c:\software\blender-4.1.1-windows-x64\4.1\python\lib\site-packages (63.2.0)
Requirement already satisfied: pip in c:\software\blender-4.1.1-windows-x64\4.1\python\lib\site-packages (23.2.1)

[notice] A new release of pip is available: 23.2.1 -> 24.2
[notice] To update, run: C:\Software\blender-4.1.1-windows-x64\4.1\python\bin\python.exe -m pip install --upgrade pip
dependencies installed
Traceback (most recent call last):
  File "C:\Users\pantze28\Pictures\install_addon.py", line 73, in <module>
    addon_zip = argv[argv.index('--') + 1]
                     ^^^^^^^^^^^^^^^^
ValueError: '--' is not in list

Blender quit
```

</p>
</details> 

When I run it with portable v4.2.1, I get:

<details><summary>Details</summary>
<p>

```
Blender 4.2.1 LTS (hash 396f546c9d82 built 2024-08-19 11:29:52)
Looking in links: c:\Users\pantze28\AppData\Local\Temp\tmpedlc_fbu
Requirement already satisfied: setuptools in c:\software\blender-4.2.1-windows-x64\4.2\python\lib\site-packages (63.2.0)
Requirement already satisfied: pip in c:\software\blender-4.2.1-windows-x64\4.2\python\lib\site-packages (23.2.1)

[notice] A new release of pip is available: 23.2.1 -> 24.2
[notice] To update, run: C:\Software\blender-4.2.1-windows-x64\4.2\python\bin\python.exe -m pip install --upgrade pip
Traceback (most recent call last):
  File "C:\Users\pantze28\Pictures\install_addon.py", line 61, in <module>
    import bidict
ModuleNotFoundError: No module named 'bidict'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\pantze28\Pictures\install_addon.py", line 66, in <module>
    raise Exception("dependency installation failed")
Exception: dependency installation failed

Blender quit
```

</p>
</details> 

When trying to install the addon from the Mastodon setup window, I get the same error as already shared in the original post.

**Comment by ChrisMzz** [comment]
I also got this error with both Blender 4.2.16 and Blender 5.0.1 on a Windows 11 computer.
I tested this on two computers (same Windows, Blender and FiJi versions) and one of them seemed to have the issue while the other didn't, so I'm not sure that it's version specific.
However I did figure out what was going wrong when debugging :

After manually running `install_addon.py`, on printing the result of 
```py
subprocess.check_output([python_path, '-m', 'pip', 'install', *packages])
```
I saw that the packages had been installed in a seemingly random Python environment elsewhere on the computer.
Somehow `python_path` points to Blender's Python distribution but pip installing still looks in `AppData/Roaming/Python` distributions? Seems like a bug to me.

--- 

Either way, by deleting the entire Python distribution that it was seeing in `AppData/Roaming/Python/Python311`, reinstalling worked properly without having to edit anything.

I also got it to work by just copying the packages from `AppData/Roaming/Python/Python311`'s `site-packages` folder over to the `site-packages` folder of Blender's Python distribution.

## TLDR

This seems to be a bug in how `subprocess.check_output([python_path, '-m', 'pip', 'install', *packages])` installs packages for the given python distribution, and is **not** a compatibility issue.

The plugin works well with Blender 4.2 and seems to work on 5.0 too from the limited testing I've done.

**Comment by stefanhahmann** [comment]
@ChrisMzz 

Thanks for your insights. It is difficult to debug this issue without being able to reproduce the error on my machine.

I gave co-pilot a try on this. It came up with a potential solution. It very much doubt, it is the correct solution, but it would be worth a try. Would you be willing to test this on your machine?

You would need to download and unzip this file: [mastodon-blender-view-0.3.5-SNAPSHOT.jar.zip](https://github.com/user-attachments/files/24562951/mastodon-blender-view-0.3.5-SNAPSHOT.jar.zip)

Then copy it to the subfolder "jar" of your Fiji installation folder and delete the mastodon-blender-view-*.jar file in it. Restart Fiji and test, if the reported error disappears.

**Comment by ChrisMzz** [comment]
@stefanhahmann 

I deleted the installed modules from Blender's local `site-packages` to make sure I had the error with `mastodon-blender-view-0.3.4.jar`, then did as you said, replacing it with `mastodon-blender-view-0.3.4-SNAPSHOT.jar`.

However, I still get the same error, which is fixed easily by adding the modules back into Blender's Python path manually.

If you can, check out my PR https://github.com/mastodon-sc/mastodon-blender-view/pull/36 and tell me if you see any destructive changes. This should ensure pip installs in Blender's install path. 
I guess maybe you can add a flag to that PR that checks if the version of Blender includes pip and chooses whether to do the changes to the install path accordingly (because it will fail to force the install path if the folders to Blender's `site-packages` don't exist).
